drivers/dma/dma_silabs_ldma.c

```diff
@@ -0,0 +1,564 @@
+		desc->xfer.decLoopCnt = 0;
+		desc->xfer.ignoreSrec = block->flow_control_mode;
+
+		/* In silabs LDMA, increment sign is managed with the transfer configuration
+		 * which is common for all descs of the channel. Zephyr DMA API allows
+		 * to manage increment sign for each block desc which can't be done with
+		 * silabs LDMA. If increment sign is different in 2 block desc, then an
+		 * error is returned.
+		 */
+		if (block->source_addr_adj != DMA_ADDR_ADJ_NO_CHANGE &&
+		    block->source_addr_adj != chan_conf->xfer_config.ldmaCfgSrcIncSign) {
+			return -ENOTSUP;
+		}
+
+		if (block->source_addr_adj == DMA_ADDR_ADJ_NO_CHANGE) {
+			desc->xfer.srcInc = ldmaCtrlSrcIncNone;
+		} else {
+			desc->xfer.srcInc = ldmaCtrlSrcIncOne;
+		}
+		if (block->dest_addr_adj == DMA_ADDR_ADJ_NO_CHANGE) {
+			desc->xfer.dstInc = ldmaCtrlDstIncNone;
+		} else {
+			desc->xfer.dstInc = ldmaCtrlDstIncOne;
+		}
+
+		desc->xfer.srcAddrMode = ldmaCtrlSrcAddrModeAbs;
+		desc->xfer.dstAddrMode = ldmaCtrlDstAddrModeAbs;
+
+		if (config->head_block->source_address == 0) {
+			LOG_WRN("source_buffer address is null.");
+		}
+
+		if (config->head_block->dest_address == 0) {
+			LOG_WRN("dest_buffer address is null.");
+		}
+
+		desc->xfer.srcAddr = block->source_address;
+		desc->xfer.dstAddr = block->dest_address;
+
+		if (block->next_block) {
+			desc->xfer.linkMode = ldmaLinkModeAbs;
+			desc->xfer.link = 1;
+
+			ret = sys_mem_blocks_alloc(data->dma_desc_pool, 1, (void **)&temp_desc);
+			if (ret) {
+				return -EXFULL;
+			}
+
+			/* Link to the next descriptor and configure it in the next iteration */
+			desc->xfer.linkAddr = LDMA_DESCRIPTOR_LINKABS_ADDR_TO_LINKADDR(temp_desc);
+			desc = temp_desc;
+			block = block->next_block;
+			/* Protection against descriptor loop*/
+			if (block == head_block) {
+				block = 0;
+			}
+		} else {
+			desc->xfer.linkMode = 0;
+			desc->xfer.link = 0;
+			desc->xfer.linkAddr = 0;
+			block = 0;
+		}
+	} while (block);
+
+	return 0;
+}
+
+static void dma_silabs_irq_handler(const struct device *dev, uint32_t id)
+{
+	const struct dma_silabs_data *data = dev->data;
+	struct dma_silabs_channel *chan;
+	int status = DMA_STATUS_COMPLETE;
+	uint32_t pending, chnum, chmask;
+
+	/* Get all pending and enabled interrupts. */
+	pending = LDMA->IF;
+	pending &= LDMA->IEN;
+
+	/* Iterate over all LDMA channels. */
+	for (chnum = 0, chmask = 1; chnum < data->dma_ctx.dma_channels; chnum++, chmask <<= 1) {
+		chan = &data->dma_chan_table[chnum];
+
+		/* Check for LDMA error. */
+		if (pending & LDMA_IF_ERROR) {
+			if (chan->cb != NULL) {
+				chan->cb(dev, chan->user_data, chnum, -EIO);
+			}
+		} else if (pending & chmask) {
+			/* Clear the interrupt flag. */
+			LDMA->IF_CLR = chmask;
+
+			/* Is it only an interrupt for the end of a descriptor and not a complete
+			 * transfer 
+			 */
+			if (!chan->complete_callback_en) {
+				status = DMA_STATUS_BLOCK;
+			} else {
+				chan->busy = false;
+			}
```

Comment:
If one channel set `DMA_STATUS_BLOCK`, all the further channels will also have `DMA_STATUS_BLOCK`. Is it expected?